Add perf measurement to the test infra, and log the results it in the DB

stevendae opened the pull request:
### Ticket
Github Issue here: https://github.com/tenstorrent/tt-metal/issues/23822

### Problem description
Perf is not being measured in sweeps infrastructure. Make this so. 

### What's changed


### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

stevendae committed:
Add performance measurement with cache support in sweeps runner

- Introduced `measure_perf_with_cache` flag in `SweepsConfig` to enable performance measurement with and without caching.
- Updated argument validation to prevent simultaneous use of `--perf` and `--perf-with-cache`.
- Enhanced performance data gathering to handle both cached and uncached results.
- Modified result collection to include separate metrics for cached and uncached performance.
- Updated logging to reflect the new performance measurement capabilities.

- [x] All tests pass
- [x] New functionality is covered by tests

stevendae committed:
re enable ability to read profiler data from device

stevendae committed:
tell profiler to immediately write csv file when readdeviceprofiler is called

stevendae committed:
Merge branch 'main' into stevendae/measure_perf

stevendae committed:
Enhance performance result handling in sweeps_runner.py and roofline_utils.py

- Improved logic for combining results from cached and uncached runs, ensuring accurate status and message reporting based on run outcomes.
- Added checks for message consistency when both runs pass, logging discrepancies as warnings.
- Simplified performance result serialization by including only essential metrics in the output queue.
- Updated message formatting in get_updated_message to handle exceptions and log processing details.

This update aims to enhance the robustness and clarity of performance reporting in the sweeps framework.

stevendae committed:
Refine result handling in sweeps_runner.py for cached and uncached runs

- Enhanced logic to combine results from cached and uncached runs, improving status and message reporting based on outcomes.
- Added detailed checks for message consistency when both runs pass, logging any discrepancies as warnings.
- Updated message formatting to provide clearer feedback on run failures and successes.

This update aims to improve the accuracy and clarity of performance reporting in the sweeps framework.

stevendae committed:
change dummy data to 0

stevendae committed:
add perf to sweeps workflow

stevendae committed:
Add TRACY_NO_INVARIANT_CHECK environment variable to sweeps workflow

stevendae committed:
use comprehensive batch size for perf runs as well.

stevendae committed:
Refine child_mode logic and update help text in sweeps_runner.py

- Clarified the conditions under which child_mode is set to True or False, improving code readability.
- Enhanced the help text for the main_proc_verbose argument to provide clearer guidance on its purpose for debugging and profiling.

Commit:
Added curated sweep tests

stevendae committed:
update timeout on conv2d short sweep

stevendae committed:
when no perf profiling datais available just use 0 for all values as dummy data

stevendae committed:
feat(sweeps): Add separate device perf capture for cached/uncached runs

When using --perf-with-cache with --device-perf flags together, the sweep
runner now captures device performance metrics separately for each run:

- First run (uncached): Captures device perf without program cache
- Second run (cached): Captures device perf with program cache benefit
- Clears profiler log between runs to prevent data accumulation

The device performance data is now structured as:
- Single run: device_perf = {metrics...}
- Cache testing: device_perf = {"uncached": {...}, "cached": {...}}

This enables accurate performance comparison between cached and uncached
execution at the device level, complementing the existing e2e performance
measurements.

stevendae committed:
feat(results): Export separate device perf metrics for cached/uncached runs

Updates FileResultDestination to properly export device performance metrics
when cache testing is enabled:

- Collects device_perf_uncached and device_perf_cached as separate metrics
- Adds _uncached/_cached suffixes to metric names for clear distinction
- Supports both flat (device_perf_uncached/cached) and nested structures
- Maintains backward compatibility with single-run device_perf field

Metrics are exported as PerfMetric objects with descriptive names:
- device_DEVICE_FW_DURATION_[ns]_uncached
- device_DEVICE_FW_DURATION_[ns]_cached
- device_DEVICE_KERNEL_DURATION_[ns]_uncached
- device_DEVICE_KERNEL_DURATION_[ns]_cached

Note: PostgreSQL and Elasticsearch destinations have been deprecated and
their special handling for device_perf fields has been removed.

stevendae committed:
remove unneeded log message

stevendae committed:
Merge branch 'stevendae/measure_perf' of github.com:tenstorrent/tt-metal into stevendae/measure_perf

stevendae committed:
sweep_framework: Fix program cache management for --perf measurements

The program cache is enabled by default when a device is initialized.
For --perf measurements, we need to disable it immediately after device
creation to ensure every test vector measures the full compilation and
execution time.

Key changes:
- Move cache disabling to immediately after device context creation
- Remove unnecessary cache re-enabling after --perf measurements
- Add clear comments explaining the cache behavior

Note: The cache is a host-side structure that stores compiled programs.
When disabled, it prevents caching of new programs and lookup of existing
cached programs.

stevendae committed:
refactor(sweeps_runner): Remove redundant program cache disabling for performance measurements

The code for disabling the program cache during performance measurements has been removed as it is no longer necessary. This change simplifies the run function in sweeps_runner.py, ensuring cleaner and more efficient execution of tests.

Additionally, minor formatting adjustments were made in roofline_utils.py for improved readability of bounds checks.

stevendae committed:
randomize test vector order

stevendae committed:
Disable end-to-end performance measurement in sweeps_runner due to kernel cache issues. Updated argument parsing to reflect this change and added comments explaining the temporary removal of performance flags. This will be revisited once kernel cache clearing functionality is available.

stevendae committed:
refactor(sweeps_runner): Correct child_mode logic in execute_suite

stevendae committed:
Write comments to indicate missing kernel cache step. e2e unsupported until implemented

stevendae committed:
remove postgres as an export destination

stevendae committed:
add validation check for sweep run batch count in gh workflow

stevendae committed:
Merge branch 'main' into stevendae/measure_perf

stevendae committed:
remove e2e related perf tracking details in ttnn-run-sweeps

stevendae committed:
Merge branch 'stevendae/measure_perf' of github.com:tenstorrent/tt-metal into stevendae/measure_perf

stevendae committed:
remove bold comment in ttnn-run-sweeps of requiring profiler build

stevendae committed:
Enhance error handling in ttnn-run-sweeps workflow by replacing exit with sys.exit for batch count validation

stevendae committed:
commit tools to create mako csvs from test vector and results jsons

stevendae committed:
Merge branch 'proberts/sweep_tests' into stevendae/measure_perf

stevendae committed:
updated get_updated_message in tests/sweep_framework/sweep_utils/roofline_utils.py to handle nested cached/uncached gracefully and minimally.

stevendae committed:
Made disable_watcher and disable_profiler idempotent with os.environ.pop(key, None) to avoid KeyError if vars are missing.

stevendae committed:
Added --shuffle-seed to sweeps_parameter_generator.py and seeded random before generation.

stevendae committed:
Added --shuffle-seed to sweeps_parameter_generator.py and seeded random before generation.

stevendae committed:
enable e2e perf capture but not default in CI

stevendae committed:
Updated result_destination.py to extract metric naming and coercion into helpers and refactored _collect_all_metrics to use them

stevendae committed:
Merge branch 'stevendae/measure_perf' of github.com:tenstorrent/tt-metal into stevendae/measure_perf

stevendae committed:
remove mako related files

stevendae committed:
remove data_movement curated

stevendae committed:
Merge branch 'main' into stevendae/measure_perf

patrickroberts reviewed (approved):
I particularly like that you log the shuffle seed when it's auto-generated.

amorrisonTT commented on tests/sweep_framework/sweeps_runner.py:
```diff
@@ -333,23 +362,164 @@ def run(test_module_name, input_queue, output_queue, config: SweepsConfig):
+
+                    # First run (without cache) - measure uncached performance
+                    results_uncached = test_module.run(**test_vector, device=device)
+                    if type(results_uncached) == list:
```
Nit:
```suggestion
                    if isinstance(results_uncached, list):
```

stevendae commented on tests/sweep_framework/sweeps_parameter_generator.py:
```diff
@@ -59,6 +60,9 @@ def export_suite_vectors_json(module_name, suite_name, vectors):
         EXPORT_DIR_PATH.mkdir()
 
+    # Randomize the order of vectors while keeping their content (and thus input_hash) unchanged
+    random.shuffle(vectors)
```
I'm keeping this as a means of randomization to remove any suspected bias due to test order.

stevendae committed:
refactor(sweeps): polish result_destination.py for idiomatic, robust Python

- Import convert_enum_values_to_strings explicitly; remove type: ignore
- Replace print calls with structured logger.info in Superset finalize
- Use timezone-aware UTC now() for run-id generation, filename timestamps, and run_end_ts
- Make run_id fallback robust (avoid None.strftime by using UTC timestamp fallback)
- Write JSON files with encoding="utf-8" while preserving atomic write pattern
- Serialize metrics deterministically by sorting on metric_name
- Correct _flatten_any_to_dotted docstring (lists preserved; no index flattening)
- Modernize typing (PEP 585 builtins: dict/list) and reorder imports per PEP 8
- Narrow exceptions: _to_float catches (TypeError, ValueError); ES validate_connection uses logger.exception for tracebacks
- Remove 'postgres' branch from ResultDestinationFactory (undefined symbol), keeping only supported destinations
- No change to ES index params (destination slated for deprecation)

stevendae committed:
Hoisted _map_status, _coerce_to_optional_string, and _collect_all_metrics to module scope, removed inner defs, and updated calls. Lints are clean.

stevendae committed:
include input_hash to skips as well

stevendae committed:
- feat(sweeps): extract cached/uncached perf logic into helpers; preserve logs and payload
  - Adds DEVICE_PERF_KEYS and helper functions to simplify run() readability
  - Keeps queue payload [status, message, e2e_perf, device_perf] and log text identical
  - Fixes minor logging variable (vector_id -> input_hash) in XFail messages

stevendae committed:
Add license to perf_util

stevendae committed:
make randomize non-default

stevendae committed:
use global naming for perf_utils

stevendae commented:
@amorrisonTT thank you for review. In light of keeping the run function readable I made a decision to extract a single run and dual run (for perf measurement of cached op execution) into a separate file. That file `/tests/sweep_framework/framework/perf_utils.py` is where the imported tests get executed and perf related metrics get captured. I believe this makes execution paths more traceable and distinguishable between the two types of modes.

stevendae commented on tests/sweep_framework/sweeps_runner.py:
```diff
@@ -858,20 +1035,31 @@ def disable_profiler():
         "--watcher", action="store_true", required=False, help="Add this flag to run sweeps with watcher enabled."
     )
+    # E2E performance measurement is inaccurate due to default kernel caching
+    # The kernel compilation cache cannot be cleared from Python, leading to misleading results
```
back story: request was made to track op execution performance under uncached and cached states. I was under the impression that only one such cache system was in place for this (program cache) and can be cleared through the mechanism you describe above. However after running sweep tests I noticed that despite clearing program cache after every dual run execution for each op + parameter set the first op execution would show low end to end performance times (as if running with the help of cache utilization). 

Looking into this I discovered kernel cache. 
https://github.com/tenstorrent/tt-metal/blob/503bd31a7a72ba68e4ea20c977c9bea6223cb615/tt_metal/impl/program/program.cpp#L1418

on cache hit it skips the `GenerateBinaries` step.

No python binding exists for this. I do have an issue created to track / implement this. https://github.com/tenstorrent/tt-metal/issues/30925

stevendae committed:
Enabled device perf on scheduled runs,
